libsigner/src/events.rs

```diff
@@ -334,47 +342,86 @@ fn process_stackerdb_event(
-            .iter()
-            .filter_map(|chunk| read_next::<SignerMessage, _>(&mut &chunk.data[..]).ok())
-            .collect();
-        SignerEvent::SignerMessages(signer_set, signer_messages)
-    } else {
-        info!(
-            "[{:?}] next_event got event from an unexpected contract id {}, return OK so other side doesn't keep sending this",
-            local_addr,
-            event.contract_id
-        );
-        if let Err(e) = request.respond(HttpResponse::empty(200u16)) {
-            error!("Failed to respond to request: {:?}", &e);
+    let event_contract_id = event.contract_id.clone();
+
+    let signer_event = match SignerEvent::try_from(event) {
+        Err(e) => {
+            info!(
+                "[{:?}] next_event got event from an unexpected contract id {}, return OK so other side doesn't keep sending this",
+                local_addr,
+                event_contract_id
+            );
+            ack_dispatcher(request);
+            return Err(e.into());
         }
-        return Err(EventError::UnrecognizedStackerDBContract(event.contract_id));
+        Ok(x) => x,
     };
 
-    if let Err(e) = request.respond(HttpResponse::empty(200u16)) {
-        error!("Failed to respond to request: {:?}", &e);
-    }
+    ack_dispatcher(request);
 
     Ok(signer_event)
 }
 
+impl TryFrom<StackerDBChunksEvent> for SignerEvent {
+    type Error = EventError;
+
+    fn try_from(event: StackerDBChunksEvent) -> Result<Self, Self::Error> {
+        let signer_event = if event.contract_id.name.as_str() == MINERS_NAME
+            && event.contract_id.issuer.1 == [0; 20]
+        {
+            let mut blocks = vec![];
+            let mut messages = vec![];
+            let mut miner_pk = None;
+            for chunk in event.modified_slots {
+                miner_pk = Some(chunk.recover_pk().map_err(|e| {
+                    EventError::MalformedRequest(format!(
+                        "Failed to recover PK from StackerDB chunk: {e}"
+                    ))
+                })?);
+                if chunk.slot_id % 2 == 0 {
+                    // block
+                    let Ok(block) =
+                        NakamotoBlock::consensus_deserialize(&mut chunk.data.as_slice())
+                    else {
+                        continue;
+                    };
+                    blocks.push(block);
+                } else if chunk.slot_id % 2 == 1 {
+                    // message
+                    let Ok(msg) = SignerMessage::consensus_deserialize(&mut chunk.data.as_slice())
+                    else {
+                        continue;
+                    };
+                    messages.push(msg);
+                } else {
+                    return Err(EventError::UnrecognizedEvent(
+                        "Unrecognized slot_id for miners contract".into(),
+                    ));
+                };
+            }
+            SignerEvent::ProposedBlocks(blocks, messages, miner_pk)
```

Comment:
The big change here is that the miners contract now submits two kinds of information: block proposals and fire coordinator messages.